enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,362 @@
+lying in this pool, so it doesn't undo or changes the allocation logic performed by the plugin.
+Therefore, the plugin assigns the shared CPUs to containers without racing/conflict
+the CPU manager behavior. 
+More about this decision at the [Alternative](mixed-cpu-node-plugin.md#alternatives) section
+
+When management and workload partitioning is enabled,
+the `reservedSystemCpus` were exclusive to management components (Kubelet/CRIO) or pods that were labeled as management. 
+With this proposal system reserved no longer maps to Kubelet's view of system reserved.
+Instead, there is an internal partition of `resrvedSystemCpus` in which the `reserved` cpus are dedicated
+and exclusive to management component ([same as today](https://github.com/openshift/cluster-node-tuning-operator/blob/master/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L592)).
+While the `shared` cpus are dedicated for workloads which are asking for shared cpus.
+
+The following shows the cpu layout and what processes can be running at what cpus:
+![cpu layout](cpu_layout.jpg)
+Please check the [Risks and Mitigations](mixed-cpu-node-plugin.md#risks-and-mitigations) for additional information.
+
+There is a plan for a feature called [shared-partition](https://github.com/openshift/enhancements/pull/1421) 
+that implements additional pool.
+When shared-partition lands, Kubelet's cpu pool layout should look roughly like:
+1. `reservedSystemCpus`
+2. `guaranteedCpus` <- new pool
+3. `sharedCpus` = All CPUs - `reservedSystemCpus` - `guaranteedCpus`
+
+When shared-partition feature enabled, 
+mixed-cpu-node-plugin allocates the shared cpus from the `sharedCpus` pool and not
+from the reservedSystemCpus anymore.
+
+The following section summarizes the shared-cpus source in each flow:
+* mixed-cpu enabled -> `cpu.shared` appends to Kubelet's `reservedSystemCPUs`
+* shared-partition enabled -> Kubelet's `sharedCpus` = All cpus - `reserved` - `isolated`
+* Both enabled -> sharedCpus = All cpus - `reserved` - `isolated`
+
+The node-plugin populates a special device named `openshift.io/shared-cpus` to provide way for pods to request
+for this special type of CPUs.
+There's no meaning to the value/number of `openshift.io/shared-cpus` devices that the pod requests.
+For example:
+```yaml
+requests:
+   openshift.io/shared-cpus: 1
+```
+Or
+```yaml
+requests:
+   openshift.io/shared-cpus: 2
+```
+Are the same in that context, the container has access to the same number of shared-cpus in both cases.
+In other words, the device request only uses as a hint to tell the plugin that shared-cpus required for the container and
+does not indicate any actual value.
+
+The reason for specifying a device, and not, for example, an annotation,
+is because when application pod requests for a device, the scheduler makes the pod pending till device up and running.
+This gives the node plugin room for setup and not being depended on pod admission order.
```

Comment:
So this will work kind of like the GPU plugins in terms of marking the node with available capacity for the scheduler? Have you considered the cluster autoscaler at all within this enhancement? It will also be affected by this enhancement

Reply:
Hey @JoelSpeed thank you for your comment.
Essentially, we're using the device request only to make sure pods are not scheduled when the plugin is not ready.
There's no actual capacity constrains here.
The node-plugin registers a dummy device during its initialization process which guarantee that no workload (which asks for shared cpus) gets admitted by the scheduler before the node-plugin is ready.
The devices are dummy generated devices, which are not associated with any actual HW.

Reply:
Ok, so the cluster autoscaler uses the same logic as the scheduler to schedule pods and then determine if it needs to be adding more capacity to the cluster. I think in this case, the node plugin, even though it's fake in a way, will affect it's scheduling decisions, and may cause either too many or too few nodes to be created, we should explore this.

CC @elmiko do you agree with my conclusion?

Reply:
Hey @JoelSpeed,
So IIUC your concern is in times when all fake devices gets allocated, and more pods are coming requesting for this shared-cpu device, the cluster autoscaler will try to spin up more nodes to overcome the shortage, am I correct?

There's a logic in the node-plugin that makes sure to always populates more devices to avoid shortage.
Everytime, the plugin counts for the allocation requests. if the number of allocation is bigger than the number of device it has, it repopulates its devices (keeping the same ids to retain the existing allocations) plus constant number of new devices (which we can make configurable if needed).
This process repeat itself everytime we'll reach maximal devices capacity.
https://github.com/openshift-kni/mixed-cpu-node-plugin/blob/main/pkg/deviceplugin/implementation.go#L67-L74 

In addition, I went through the [docs](https://docs.openshift.com/container-platform/4.13/machine_management/applying-autoscaling.html) and it seems that the cluster autoscaler design to watch for cpu, memory and GPU devices only (optional) 
so it doesn't watch for the customized shared-cpu devices.

Reply:
@JoelSpeed good thinking, and in general i agree but i would like to understand the issue here a little deeper.

in general, if we have a controller that will change a node's resources (capacity or allocatable) or its taints and labels, then we potentially have a situation where the cluster autoscaler could get confused about what is in the cluster.

we see this most frequently with gpu usage. essentially, when the autoscaler is creating a new node that will eventually contain a gpu (after drivers are installed), then it needs to know that a gpu will be there so that it doesn't make too many nodes. i have a feeling we could encounter something similar in the scenario described in this thread. 

the autoscaler needs to know that "this node will eventually have resource X, but does not currently". there is a mechanism for this in the autoscaler, but it's also an area of development that needs more attention.

Reply:
Hey @elmiko is the cluster autoscaler watches for every type of resources on the node, including customize device as the shared-cpu device?

And if the answer is yes, is it possible to configure the cluster autoscaler to ignore certain type of device?

From the docs it seems the configuration is a possibility

Reply:
@Tal-or the autoscaler doesn't necessarily watch for specific resources by default, but if there are workloads which request a resource then the autoscaler will attempt to find nodes that can satisfy that resource. if it does find those nodes, then it will try to make more of those nodes.

i think the concern here is that we would have some workload which requests a resource that requires configuration before it appears on a node. in these cases, the autoscaler will create a new node, see the new node, and see that the workload is pending until the resource is available. *but*, since the autoscaler doesn't know that the resource will appear later, it creates another node. this pattern continues until workloads start getting scheduled, by which time too many nodes have been created.

Reply:
Valid concern indeed.

> i think the concern here is that we would have some workload which requests a resource that requires configuration before it appears on a node. 

This scenario can only happens when node gets rebooted, Kubelet starts all the pods which were presents at the node prior to the reboot, and the workload gets admitted by Kubelet before the node plugin. 

The other scenario I can think of is when user creates a workload which asks for shared CPU device, but never enabled this feature.
But this problem is true for every workload which asks for devices that doesn't exists (might be even a typo in the device name under the pod spec request)

So how the autoscaler deals with such scenarios today?

Reply:
> The other scenario I can think of is when user creates a workload which asks for shared CPU device, but never enabled this feature.

would we ever have a scenario where a workload has a request for a shared cpu device and we would have to wait for something to add that resource to an existing node? (eg a new node has become ready and must wait for a driver to enable the shared cpu)

> But this problem is true for every workload which asks for devices that doesn't exists (might be even a typo in the device name under the pod spec request)

true

> So how the autoscaler deals with such scenarios today?

it really depends on the scenario, but lets say that i create a Pod and set the requests value to `gpy: 1` (i meant to type "gpu" but flubbed it). in this case the autoscaler will see that it cannot make a node which supplies the requested `qpy` resource and would emit warnings in the log that it has no node groups to satisfy the pending pod.

the problem isn't when the resource does not exist at all, the problem arises when the resource does exist but takes time to install after the node becomes ready.